### PR TITLE
fuse: add benchmarking support for FUSE

### PR DESCRIPTION
**Pull request by ridwanmsharif**
This change adds the following:
- [x] Add support for containerizing syscall tests for FUSE
- [x] Mount tmpfs in the container so we can run benchmarks against it
- [x] Run the server in a background process
- [x] Implement syscalls and collect result

Steps to run:
Make sure you have VFS2 and FUSE enabled on your runsc.

`make load-all-images`
`bazel test --test_output=all --cache_test_results=no --test_arg=-test.bench= //test/fuse:stat_benchmark_runsc_ptrace_vfs2_fuse_container` 

To add benchmarks, see test/perf/stat_benchmark for inspiration and see the BUILD
file to see how to set the configs to use a FUSE container.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -334,7 +334,13 @@ RUNTIME_BIN     := $(RUNTIME_DIR)/runsc
 RUNTIME_LOG_DIR := $(RUNTIME_DIR)/logs
 RUNTIME_LOGS    := $(RUNTIME_LOG_DIR)/runsc.log.%TEST%.%TIMESTAMP%.%COMMAND%
 
+ifeq (,$(RUNTIME_NAME))
+RUNTIME_NAME := $(RUNTIME)
+endif
+
 dev: ## Installs a set of local runtimes. Requires sudo.
+
+	echo "HELLO"
```
Comment (jinmouil) [inline Makefile]: Is this for debugging or we need this in the production code?

File: images/basic/fuse/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+RUN apt update
+RUN add-apt-repository ppa:ubuntu-toolchain-r/test
+RUN apt-get install -y build-essential git pkg-config fuse3 libfuse3-3 libfuse3-dev strace
+RUN apt-get install -y procps vim libstdc++6
+
+WORKDIR /fus
+
+RUN mkdir -pv mountpoint
+
+RUN git clone https://github.com/libfuse/libfuse
```
Comment (jinmouil) [inline images/basic/fuse/Dockerfile:11]: There 2 `git clone` may not be required

**Comment by ridwanmsharif** [comment]
This PR isn't quite ready for review yet. When it is, we'll make it a non-draft PR.

File: images/basic/fuse/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    zlib1g-dev \
+    unzip \
+    python3 \
+    && rm -rf /var/lib/apt/lists/*
+
+RUN apt-get update && apt-get install -y software-properties-common
+
+RUN apt update
+RUN add-apt-repository ppa:ubuntu-toolchain-r/test
+RUN apt-get install -y build-essential git pkg-config fuse3 libfuse3-3 libfuse3-dev strace
```
Comment (jinmouil) [inline images/basic/fuse/Dockerfile:6]: Why install git multiple times? And `apt update` multiple time?

File: pkg/sentry/fs/g3doc/fuse.md [hunk 4]
```diff
@@ -254,6 +254,51 @@ I/O syscalls like `read(2)`, `write(2)` and `mmap(2)`.
 -   `FUSE_BMAP`: Old address space API for block defrag. Probably not needed.
 -   `FUSE_NOTIFY_REPLY`: [TODO: what does this do?]
 
+## Benchmark FUSE
+
+FUSE benchmark makes FUSE syscall inside docker container to make sure required environment are met.
```
Comment (jinmouil) [inline pkg/sentry/fs/g3doc/fuse.md]: "required environment are met"?

File: test/runner/runner.go [hunk 5]
```diff
@@ -399,6 +489,7 @@ func runTestCaseRunsc(testBin string, tc gtest.TestCase, t *testing.T) {
 
 	if *addUDSTree {
 		cleanup, err := setupUDSTree(spec)
+		_ = cleanup
```
Comment (jinmouil) [inline test/runner/runner.go]: ?

File: test/runner/runner.go [hunk 6]
```diff
@@ -256,7 +262,7 @@ func runRunsc(tc gtest.TestCase, spec *specs.Spec) error {
 	if err == nil && len(testLogDir) > 0 {
 		// If the test passed, then we erase the log directory. This speeds up
 		// uploading logs in continuous integration & saves on disk space.
-		os.RemoveAll(testLogDir)
+		//os.RemoveAll(testLogDir)
```
Comment (jinmouil) [inline test/runner/runner.go]: Do we need to change all the gVisor test?

**Review by jinmouil — CHANGES_REQUESTED** [review]

File: images/basic/fuse/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+RUN apt-get install -y procps vim libstdc++6
+
+WORKDIR /fus
+
+RUN mkdir -pv mountpoint
+COPY server-bin ./server-bin
```
Comment (craig08) [inline images/basic/fuse/Dockerfile]: Why don't we build binary in this container? Pushing a binary file in git source tree is usually not a good idea.

File: images/basic/fuse/Dockerfile (hunk 3 above)
Comment (craig08) [reply images/basic/fuse/Dockerfile:6]: This Dockerfile should be cleaned up. There are repeated and unused packages.

File: pkg/test/dockerutil/container.go [hunk 8]
```diff
@@ -397,7 +402,7 @@ func (c *Container) CopyFiles(opts *RunOpts, target string, sources ...string) {
 		c.copyErr = fmt.Errorf("ioutil.TempDir failed: %v", err)
 		return
 	}
-	c.cleanups = append(c.cleanups, func() { os.RemoveAll(dir) })
+	//c.cleanups = append(c.cleanups, func() { os.RemoveAll(dir) })
```
Comment (craig08) [inline pkg/test/dockerutil/container.go]: Why do we remove cleanup here?

Comment (ridwanmsharif) [reply pkg/test/dockerutil/container.go]: Oops, come of this was for when I was debugging. Fixed now.

File: images/basic/fuse/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+# Get everything needed for a FUSE Server.
+RUN add-apt-repository ppa:ubuntu-toolchain-r/test
+RUN apt-get install -y build-essential fuse3 libfuse3-3 libfuse3-dev strace
+RUN apt-get install -y procps vim libstdc++6
+
+WORKDIR /fus
+
+RUN mkdir -pv mountpoint
+RUN git clone https://github.com/libfuse/libfuse
+RUN gcc -Wall ./libfuse/example/hello.c `pkg-config fuse3 --cflags --libs` -o server-bin
```
Comment (craig08) [inline images/basic/fuse/Dockerfile]: This should be `passthrough.c`. I think what the image does is to build a `passthrough` and enable fuse library. Do you think the following `Dockerfile` work? Why do we need software-properties-common and add toolchain repo?
```
FROM ubuntu:20.04

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update
RUN apt-get install -y build-essential git pkg-config fuse3 libfuse3-3 libfuse3-dev strace

WORKDIR /fus

RUN git clone https://github.com/libfuse/libfuse
RUN mkdir -pv mountpoint

RUN gcc -Wall ./libfuse/example/passthrough.c `pkg-config fuse3 --cflags --libs` -o server-bin

CMD ["bash"]
```

File: test/perf/BUILD [hunk 10]
```diff
@@ -55,6 +55,7 @@ syscall_test(
     size = "large",
     add_overlay = True,
     test = "//test/perf/linux:open_benchmark",
+    vfs2 = "True",
```
Comment (craig08) [inline test/perf/BUILD]: Isn't vfs2 default enabled?

File: test/runner/defs.bzl [hunk 11]
```diff
@@ -142,6 +150,7 @@ def syscall_test(
       shard_count: shards for defined tests.
       size: the defined test size.
       use_tmpfs: use tmpfs in the defined tests.
+      use_fuse: use fuse in the defined tests.
```
Comment (craig08) [inline test/runner/defs.bzl]: `use_image` and `setup_command`

File: test/runner/defs.bzl [hunk 12]
```diff
@@ -184,6 +212,8 @@ def syscall_test(
         size = size,
         platform = "native",
         use_tmpfs = False,
+        use_image = "",
+        setup_command = "",
```
Comment (craig08) [inline test/runner/defs.bzl:216]: We already have default arguments, don't need this.

File: test/runner/defs.bzl [hunk 13]
```diff
@@ -195,6 +225,8 @@ def syscall_test(
             size = size,
             platform = platform,
             use_tmpfs = use_tmpfs,
+            use_image = "",
+            setup_command = "",
```
Comment (craig08) [inline test/runner/defs.bzl:229]: We already have default arguments, don't need this.

File: test/runner/defs.bzl [hunk 14]
```diff
@@ -207,6 +239,8 @@ def syscall_test(
             size = size,
             platform = default_platform,
             use_tmpfs = use_tmpfs,
+            use_image = "",
+            setup_command = "",
```
Comment (craig08) [inline test/runner/defs.bzl:243]: We already have default arguments, don't need this.

File: test/runner/defs.bzl [hunk 15]
```diff
@@ -219,6 +253,8 @@ def syscall_test(
...
```
Comment (craig08) [inline test/runner/defs.bzl:257]: We already have default arguments, don't need this.

File: test/runner/defs.bzl [hunk 16]
```diff
@@ -232,6 +268,8 @@ def syscall_test(
...
```
Comment (craig08) [inline test/runner/defs.bzl:272]: We already have default arguments, don't need this.

File: test/runner/defs.bzl (hunk 12 above)
Comment (craig08) [inline test/runner/defs.bzl]: Why don't we add default arguments to `_syscall_test` then we don't have to add these 2 variables everywhere?

File: test/runner/runner.go [hunk 17]
```diff
@@ -310,8 +316,92 @@ func setupUDSTree(spec *specs.Spec) (cleanup func(), err error) {
...
+		// Mount a tmpfs directory to use when benchmarking.
+		Mounts: []mount.Mount{
+			{
+				Type:     mount.TypeTmpfs,
+				Target:   tmpDir,
+				ReadOnly: false,
+			},
+		},
+		Env: []string{
+			fmt.Sprintf("TEST_TMPDIR=%s", "/tmpDir"),
```
Comment (craig08) [inline test/runner/runner.go]: `tmpDir` instead of `"/tmpDir"`

Comment (ridwanmsharif) [reply test/runner/runner.go]: It has to be an absolute path.

File: test/image/image_test.go [hunk 18]
```diff
@@ -63,6 +64,57 @@ func TestHelloWorld(t *testing.T) {
...
+
+		// Mount a tmpfs directory to use when benchmarking.
+		Mounts: []mount.Mount{
+			{
+				Type:     mount.TypeTmpfs,
+				Target:   tmpDir,
+				ReadOnly: false,
+			},
+		},
+	}, "sleep", "1000")
```
Comment (craig08) [inline test/image/image_test.go:94]: nit: I'm not sure what's the best practice of keeping a container alive. I learned from Jenkins to use `docker run -it ubuntu cat` to make a container alive forever. I think it's fine using sleep.

File: test/perf/BUILD (hunk 10 above)
Comment (ridwanmsharif) [reply test/perf/BUILD]: Nope, its only used when the test specifies it.

File: test/runner/runner.go [hunk 19]
```diff
@@ -310,8 +316,92 @@ func setupUDSTree(spec *specs.Spec) (cleanup func(), err error) {
...
+
+	cmd = containerTestBin + " " + strings.Join(tc.Args(), " ")
+	out, err = d.Exec(ctx, dockerutil.ExecOpts{
+		Privileged: true,
+	}, "/bin/sh", "-c", cmd)
+	if err != nil {
+		t.Fatalf("docker exec failed: %v with output %v", err, out)
+	}
+
+	fmt.Print(out)
```
Comment (craig08) [inline test/runner/runner.go:394]: Do we really print something here?

File: test/perf/BUILD (hunk 10 above)
Comment (craig08) [reply test/perf/BUILD]: I saw in `test/runner/defs.bzl` around line 143 it is default true. If it is not, why it is only added in open_benchmark? Other tests don't need vfs2?

File: test/runner/runner.go (hunk 17 above)
Comment (craig08) [reply test/runner/runner.go]: That's right, you have a variable with this absolute path in line 331.

File: images/basic/fuse/Dockerfile [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get update
+RUN apt-get install -y build-essential git pkg-config fuse3 libfuse3-3 libfuse3-dev strace
+
+WORKDIR /fus
+
+RUN mkdir -pv mountpoint
+RUN git clone https://github.com/libfuse/libfuse
```
Comment (jinmouil) [inline images/basic/fuse/Dockerfile:11]: `RUN gcc -D_FILE_OFFSET_BITS=64 -Wall /usr/share/doc/libfuse3-dev/examples/passthrough.c `pkg-config fuse3 --cflags --libs` -o passthrough`

Can we avoid this git clone like this?

File: test/perf/BUILD (hunk 10 above)
Comment (ridwanmsharif) [reply test/perf/BUILD]: Aaaaahhhh I see what happened, just a few days ago all the syscall tests made vfs2 the default. This PR started before that was true, updated it now.

File: pkg/sentry/fs/g3doc/fuse.md [hunk 21]
```diff
@@ -254,6 +254,52 @@ I/O syscalls like `read(2)`, `write(2)` and `mmap(2)`.
 -   `FUSE_BMAP`: Old address space API for block defrag. Probably not needed.
 -   `FUSE_NOTIFY_REPLY`: [TODO: what does this do?]
 
+## Benchmark FUSE
+
+FUSE benchmark makes FUSE syscall inside docker container to make sure required environment conditions
+are met - such as having the right libraries to start a FUSE server.
+
+### Setup
+To run benchmark
```
Comment (jinmouil) [inline pkg/sentry/fs/g3doc/fuse.md]: Suggest newline around code blocks and lists

File: test/image/image_test.go [hunk 22]
```diff
@@ -63,6 +64,57 @@ func TestHelloWorld(t *testing.T) {
 	}
 }
 
+// Test that the FUSE container is set up and being used properly.
+func TestFUSEInContainer(t *testing.T) {
+	if usingFUSE, err := dockerutil.UsingFUSE(); !usingFUSE {
```
Comment (craig08) [inline test/image/image_test.go:69]: nit: should we check `err` before `usingFUSE`?

**Review by craig08 — APPROVED** [review]
LGTM

File: images/basic/fuse/Dockerfile (hunk 20 above)
Comment (boyuan-he) [reply images/basic/fuse/Dockerfile:11]: The command doesn't work for me, `gcc: error: /usr/share/doc/libfuse3-dev/examples/passthrough.c`, further investigation shows `/usr/share/doc/libfuse3-dev/examples` is empty

**Pull request merged**